core/services/wasi-fs/src/core.rs

```diff
@@ -0,0 +1,247 @@
+        let mode = match stat.type_ {
+            DescriptorType::Directory => EntryMode::DIR,
+            DescriptorType::RegularFile => EntryMode::FILE,
+            _ => EntryMode::Unknown,
+        };
+
+        let mut metadata = Metadata::new(mode).with_content_length(stat.size);
+
+        if let Some(mtime) = stat.data_modification_timestamp {
+            // Convert WASI timestamp (seconds + nanoseconds) to Timestamp
+            if let Ok(ts) = Timestamp::new(mtime.seconds as i64, mtime.nanoseconds as i32) {
+                metadata = metadata.with_last_modified(ts);
+            }
+        }
+
+        metadata
+    }
+
+    pub fn create_dir(&self, path: &str) -> Result<()> {
+        let abs_path = self.build_path(path);
+        self.root_descriptor
+            .create_directory_at(&abs_path)
+            .map_err(parse_wasi_error)
+    }
+
+    pub fn open_file(
+        &self,
+        path: &str,
+        flags: OpenFlags,
+        desc_flags: DescriptorFlags,
+    ) -> Result<Descriptor> {
+        let abs_path = self.build_path(path);
+        self.root_descriptor
+            .open_at(PathFlags::empty(), &abs_path, flags, desc_flags)
+            .map_err(parse_wasi_error)
+    }
+
+    pub fn read_dir(&self, path: &str) -> Result<DirectoryEntryStream> {
+        let abs_path = self.build_path(path);
+
+        let dir_desc = if abs_path.is_empty() {
+            // For the root, we need to open the same directory again
+            // since we can't clone the descriptor
+            self.root_descriptor
+                .read_directory()
+                .map_err(parse_wasi_error)?;
+            return self
+                .root_descriptor
+                .read_directory()
+                .map_err(parse_wasi_error);
+        } else {
+            self.root_descriptor
+                .open_at(
+                    PathFlags::empty(),
+                    &abs_path,
+                    OpenFlags::DIRECTORY,
+                    DescriptorFlags::empty(),
+                )
+                .map_err(parse_wasi_error)?
+        };
+
+        dir_desc.read_directory().map_err(parse_wasi_error)
+    }
+
+    pub fn delete_file(&self, path: &str) -> Result<()> {
+        let abs_path = self.build_path(path);
+        self.root_descriptor
+            .unlink_file_at(&abs_path)
+            .map_err(parse_wasi_error)
+    }
+
+    pub fn delete_dir(&self, path: &str) -> Result<()> {
+        let abs_path = self.build_path(path);
+        self.root_descriptor
+            .remove_directory_at(&abs_path)
+            .map_err(parse_wasi_error)
+    }
+
+    pub fn rename(&self, from: &str, to: &str) -> Result<()> {
+        let from_path = self.build_path(from);
+        let to_path = self.build_path(to);
+        self.root_descriptor
+            .rename_at(&from_path, &self.root_descriptor, &to_path)
+            .map_err(parse_wasi_error)
+    }
+
+    pub fn copy(&self, from: &str, to: &str) -> Result<()> {
```

Comment:
We don't need to manually simulate copying if it lacks native copy support.

Reply:
OK I'll remove the custom copy impl.

Reply:
> OK I'll remove the custom copy impl.

Just to add a bit more context:

  1. WASI doesn't have native copy (no `copy_file_range` equivalent)
  2. OpenDAL doesn't provide a copy fallback - if a backend doesn't support copy, users get an error
  3. The `fs` service uses `tokio::fs::copy` which delegates to the OS (which *may* use `copy_file_range` on Linux or a read/write loop internally - but that's the OS's responsibility, not OpenDAL's)

IMHO, if point 3 stands for `fs`, it does sort of stand for `wasi` too. But the lack of customizable buffer size is a problem all by itself. Which is a good enough a reason to keep this a user level job and remove `copy()`.